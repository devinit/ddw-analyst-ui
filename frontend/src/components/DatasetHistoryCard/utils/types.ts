export interface SavedQueryData {
  id: number;
  saved_query_db_table: string;
  operation: number;
  full_query: string | null;
  active: boolean;
  description: string;
  created_on: string;
  user: string;
  status: string;
}
