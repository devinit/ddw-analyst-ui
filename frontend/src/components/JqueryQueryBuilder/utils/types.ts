export type JqueryQueryBuilderFilter = {
  condition?: string;
  rules?: (JqueryQueryBuilderFilterComparator | JqueryQueryBuilderFilter)[];
};

export type JqueryQueryBuilderHaving = {
  condition?: string;
  rules?: (JqueryQueryBuilderHavingComparator | JqueryQueryBuilderHaving)[];
};

export type JqueryQueryBuilderHavingComparator = {
  id: string;
  field: string;
  type: string;
  input: string;
  operator: JqueryQueryBuilderComps;
  value?: string | number | { plain: string | number } | { column: string; aggregate: string };
  values?: { value: string; label: string }[];
};

export type JqueryQueryBuilderFilterComparator = {
  id: string;
  field: string;
  type: string;
  input: string;
  operator: JqueryQueryBuilderComps;
  value: string | number;
};
export type JqueryQueryBuilderComps =
  | 'equal'
  | 'not_equal'
  | 'less'
  | 'less_or_equal'
  | 'greater'
  | 'greater_or_equal'
  | 'between'
  | 'in';

export type JqueryQueryBuilderFieldData = {
  id: string | number;
  label: string;
  type: string;
};

export type JqueryQueryBuilderIcons = {
  add_group: string;
  add_rule: string;
  remove_group: string;
  remove_rule: string;
  error: string;
};
