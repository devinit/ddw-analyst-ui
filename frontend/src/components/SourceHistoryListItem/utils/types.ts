export interface FrozenData {
  id: number;
  parent_db_table: string;
  frozen_db_table: string | null;
  active: boolean;
  description: string;
  created_on: string;
  user: string;
  status: string;
  logs: string | null;
}
