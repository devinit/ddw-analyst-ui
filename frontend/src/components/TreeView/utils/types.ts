import { ReactElement } from 'react';

export interface TreeViewProps {
  data: Data[];
  depth?: number;

  deleteElement?: ReactElement;

  getStyleClassCb?: (node: Data) => void;

  isCheckable?: (node: Data, depth?: number) => void;
  isDeletable?: (node: Data, depth?: number) => void;
  isExpandable?: (node: Data, depth?: number) => void;

  keywordChildren?: string;
  keywordChildrenLoading?: string;
  keywordKey?: string;
  keywordLabel?: string;

  loadingElement?: ReactElement;
  noChildrenAvailableMessage?: string;

  onCheckToggleCb?: (data: Data[], depth?: number) => void;
  onDeleteCb?: (node: Data, data: Data[], depth?: number) => void;
  onExpandToggleCb?: (node: Data, depth?: number) => void;
  onUpdateCb?: (data: unknown[], depth?: number) => void;

  transitionEnterTimeout?: number;
  transitionExitTimeout?: number;
}

export interface Data {
  id: number;
  name: string;
  isExpanded?: boolean;
  isChecked?: boolean;
  isChildrenLoading?: boolean;
  children?: Data[];
}
