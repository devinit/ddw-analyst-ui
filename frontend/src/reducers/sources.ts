import { List, Map, fromJS } from 'immutable';
import { Action, Reducer } from 'redux';

export type SourcesAction = Action & { sources?: Source[], activeSourceIndex?: number };

export interface Source {
  pk: number;
  indicator: string;
  indicator_acronym: string | null;
  source: string | null;
  source_acronym: string | null;
  source_url: string | null;
  download_path: string | null;
  last_updated_on: string;
  storage_type: string | null;
  active_mirror_name: string | null;
  description: string | null;
  created_on: string;
  updated_on: string | null;
  sourcecolumnmap_set: Columns;
  updatehistory_set: UpdateHistory;
}
export interface Columns {
  pk: number;
  name: string | null;
  description: string | null;
  source_name: string | null;
}
export interface UpdateHistory {
  source: number;
  history_table: string | null;
  is_major_release: boolean;
  released_on: string | null;
  release_description: string | null;
  invalidated_on: string | null;
  invalidation_description: string | null;
}
interface State {
  loading: boolean;
  sources: List<SourceMap>;
  activeSourceIndex: number;
}
export type SourceMap = Map<keyof Source, Source[keyof Source]>;
export type SourcesState = Map<keyof State, State[keyof State]>;

const prefix = 'sources';
export const FETCH_SOURCES = `${prefix}.FETCH`;
export const FETCH_SOURCES_SUCCESSFUL = `${prefix}.FETCH_SUCCESSFUL`;
export const FETCH_SOURCES_FAILED = `${prefix}.FETCH_FAILED`;
export const SET_ACTIVE_SOURCE_INDEX = `${prefix}.SET_ACTIVE_SOURCE_INDEX`;

const defaultState: SourcesState = fromJS({ loading: false, sources: [], activeSourceId: 1 });

export const sourcesReducer: Reducer<SourcesState, SourcesAction> = (state = defaultState, action) => {
    if (action.type === FETCH_SOURCES) {
      return state.set('loading', true);
    }
    if (action.type === FETCH_SOURCES_SUCCESSFUL && action.sources) {
      const sources = fromJS(action.sources);
      const activeSourceIndex = action.sources.length ? 0 : -1;

      return state.withMutations(map =>
        map.set('loading', false).set('sources', sources).set('activeSourceIndex', activeSourceIndex)
      );
    }
    if (action.type === FETCH_SOURCES_FAILED) {
      return state.set('loading', false);
    }
    if (action.type === SET_ACTIVE_SOURCE_INDEX && typeof action.activeSourceIndex === 'number') {
      return state.set('activeSourceIndex', action.activeSourceIndex);
    }

    return state;
};
