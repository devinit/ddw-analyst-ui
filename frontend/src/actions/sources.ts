import { Action } from 'redux';
import { FETCH_SOURCE, FETCH_SOURCES, SET_SOURCE, SourcesAction } from '../reducers/sources';
import { SourceMap } from '../types/sources';

export const fetchSources = (): Action => ({ type: FETCH_SOURCES });
export const fetchActiveSource = (id: string | number) => ({ type: FETCH_SOURCE, payload: { id } });
export const setActiveSource = (activeSource: SourceMap, loading = false): Partial<SourcesAction> =>
  ({ type: SET_SOURCE, activeSource, loading });
