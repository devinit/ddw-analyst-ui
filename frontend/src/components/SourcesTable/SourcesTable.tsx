import classNames from 'classnames';
import { List } from 'immutable';
import { debounce } from 'lodash';
import * as React from 'react';
import { Card, FormControl, Table } from 'react-bootstrap';
import { SourceMap } from '../../reducers/sources';
import { SourcesTableRow } from '../SourcesTableRow';

interface SourcesTableProps {
  sources: List<SourceMap>;
  activeSource?: SourceMap;
  onRowClick: (source: SourceMap) => void;
}

interface SourcesTableState {
  sources: List<SourceMap>;
  searchQuery: string;
}

export class SourcesTable extends React.Component<SourcesTableProps, SourcesTableState> {
  state: SourcesTableState = {
    sources: this.props.sources,
    searchQuery: ''
  };

  static querySources(sources: List<SourceMap>, searchQuery: string) {
    return sources.filter(source => {
      const inIndicator = (source.get('indicator') as string).toLowerCase().indexOf(searchQuery) > -1;
      const inAcronym = (source.get('indicator_acronym') as string).toLowerCase().indexOf(searchQuery) > -1;

      return inIndicator || inAcronym;
    });
  }

  static getDerivedStateFromProps(props: SourcesTableProps, state: SourcesTableState): SourcesTableState | null {
    if (state.searchQuery) {
      const sources = SourcesTable.querySources(props.sources, state.searchQuery);
      if (sources.count()) {
        props.onRowClick(sources.get(0) as SourceMap);
      }

      return {
        sources,
        searchQuery: state.searchQuery
      };
    } else {
      const count = props.sources.count();
      if (count !== state.sources.count()) {
        props.onRowClick(props.sources.get(0) as SourceMap);

        return {
          sources: props.sources,
          searchQuery: state.searchQuery
        };
      }
    }

    return null;
  }

  render() {
    return (
      <Card>
        <Card.Body>
          <Card.Header>
            <FormControl
              placeholder="Search ..."
              className="w-50"
              onChange={ debounce(this.onSearchChange, 1000, { leading: true }) }
            />
          </Card.Header>
          <Table responsive hover striped>
            <thead>
              <tr>
                <th>#</th>
                <th>Indicator</th>
                <th>Updated On</th>
              </tr>
            </thead>
            <tbody>
              { this.renderRows(this.state.sources, this.props.activeSource) }
            </tbody>
          </Table>
        </Card.Body>
      </Card>
    );
  }

  private renderRows(sources: List<SourceMap>, activeSource?: SourceMap) {
    if (sources && sources.size && activeSource) {
      return sources.map((source, index) => (
        <SourcesTableRow
          key={ index }
          count={ index + 1 }
          classNames={ classNames({ 'table-danger':  activeSource.get('pk') === source.get('pk') }) }
          onClick={ () => this.props.onRowClick(source) }
          indicator={ source.get('indicator') as string }
          indicatorAcronym={ source.get('indicator_acronym') as string }
          updatedOn={ source.get('last_updated_on') as string }
        />
      ));
    }

    return null;
  }

  private onSearchChange = (event: React.FormEvent<any>) => {
    const { value: searchQuery } = event.currentTarget as HTMLInputElement;
    if (searchQuery) {
      this.setState({ searchQuery });
    } else {
      this.setState({ searchQuery: '' });
    }
  }
}
