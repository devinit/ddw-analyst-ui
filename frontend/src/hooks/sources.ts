import axios, { AxiosResponse } from 'axios';
import { fromJS, List } from 'immutable';
import * as localForage from 'localforage';
import { useEffect, useState } from 'react';
import { APIResponse } from '../types/api';
import { Source, SourceMap } from '../types/sources';
import { api, localForageKeys } from '../utils';
interface Options {
  limit: number;
  offset: number;
  link?: string;
  search?: string;
  id?: number | string;
}

const defaultOptions: Options = {
  limit: 10,
  offset: 0,
  link: '',
  search: '',
};

export const useSources = (options: Options = defaultOptions): List<SourceMap> => {
  const [token, setToken] = useState('');
  const [sources, setSources] = useState<List<SourceMap>>(fromJS([]));
  useEffect(() => {
    localForage.getItem<string>(localForageKeys.API_KEY).then((_token) => setToken(_token));
  }, []);
  useEffect(() => {
    if (!token) {
      return;
    }
    const url = `${api.routes.SOURCES}?limit=${options.limit}&offset=${options.offset}&search=${
      options.search || ''
    }`;
    axios
      .request({
        url: options.link || url,
        method: 'get',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `token ${token}`,
        },
      })
      .then(({ status, data, statusText }: AxiosResponse<APIResponse<Source[]>>) => {
        if (status === 200 && data.results) {
          setSources(fromJS(data.results));
        } else if (status === 401) {
          console.log('Failed to fetch sources: ', statusText);
          setSources(fromJS([]));
        }
      })
      .catch((error) => {
        console.log(
          `Failed to fetch sources: ${error.response.status} ${error.response.statusText}: ${error.response.data.detail}`,
        );
        setSources(fromJS([]));
      });
  }, [token, options.limit, options.offset, options.search]);

  return sources;
};

interface UseSourceResult {
  source?: SourceMap;
  loading: boolean;
}

export const useSource = (id: number, fetch = false): UseSourceResult => {
  const [source, setSource] = useState<SourceMap | undefined>(fromJS({}));
  const [loading, setLoading] = useState(false);

  const fetchSource = () => {
    setLoading(true);
    axios
      .request({
        url: `${api.routes.SOURCES}${id}`,
        method: 'get',
        headers: {
          'Content-Type': 'application/json',
        },
      })
      .then(({ status, data, statusText }: AxiosResponse<Source>) => {
        if (status === 200 && data) {
          const activeSource = fromJS(data);
          setSource(activeSource);
          setLoading(false);
        } else if (status === 401) {
          console.log('Failed to fetch source: ', statusText);
          setSource(undefined);
          setLoading(false);
        }
      })
      .catch((error) => {
        console.log(
          `Failed to fetch source: ${error.response.status} ${error.response.statusText}`,
        );
        setSource(undefined);
        setLoading(false);
      });
  };

  useEffect(() => {
    if (fetch) {
      fetchSource();
    } else {
      localForage
        .getItem<Source | undefined>(localForageKeys.ACTIVE_SOURCE)
        .then((activeSource) => {
          if (activeSource && activeSource.id === id) {
            setSource(fromJS(activeSource));
            setLoading(false);
          } else {
            fetchSource();
          }
        });
    }
  }, [id]);

  return { loading, source };
};
