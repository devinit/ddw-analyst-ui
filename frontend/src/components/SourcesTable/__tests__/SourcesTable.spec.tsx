/**
 * @jest-environment jsdom
 */
import { List, Map } from 'immutable';
import * as React from 'react';
import { cleanup, render } from 'react-testing-library';
import * as TestRenderer from 'react-test-renderer';
import { ColumnList, SourceMap, UpdateHistoryList } from '../../../types/sources';
import { SourcesTable } from '../SourcesTable';

const columns: ColumnList = List([ Map({ source_name: 'Column 1', description: 'Column One' }) ]) as ColumnList;
const updateHistory: UpdateHistoryList = List([
  Map({ released_on: new Date('August 19, 2018 23:15:30').toISOString(), release_description: 'My Release' })
]) as UpdateHistoryList;
const source: SourceMap = Map({
  id: 1,
  indicator: 'Common Reporting Standard',
  indicator_acronym: 'crs',
  columns,
  update_history: updateHistory,
  source_acronym: 'oecd',
  last_updated_on: new Date('August 19, 2018 23:15:30').toISOString()
}) as SourceMap;
const source2 = source
  .set('id', 2)
  .set('indicator', 'World Bank Indicators')
  .set('indicator_acronym', 'wdi')
  .set('last_updated_on', new Date('August 19, 2018 23:15:30').toISOString());
let sourcesList: List<SourceMap> = List([ source, source2 ]);

afterEach(cleanup);

test('renders correctly with default props', () => {
  const renderer = TestRenderer
    .create(<SourcesTable activeSource={ source } sources={ sourcesList } onRowClick={ jest.fn() }/>)
    .toJSON();

  expect(renderer).toMatchSnapshot();
});

test('renders an empty table if no sources are provided', () => {
  const renderer = TestRenderer
    .create(<SourcesTable activeSource={ source } sources={ List() } onRowClick={ jest.fn() }/>)
    .toJSON();

  expect(renderer).toMatchSnapshot();
});

test('renders correctly when the sources are updated', () => {
  const { container, rerender } = render(
    <SourcesTable activeSource={ source } sources={ sourcesList } onRowClick={ jest.fn() }/>
  );
  sourcesList = sourcesList.push(source
    .set('id', 3)
    .set('indicator', 'Official Development Assistance')
    .set('indicator_acronym', 'DAC1')
    .set('last_updated_on', new Date('August 19, 2018 23:15:30').toISOString()));
  rerender(<SourcesTable activeSource={ source } sources={ sourcesList } onRowClick={ jest.fn() }/>);

  expect(container).toMatchSnapshot();
});
