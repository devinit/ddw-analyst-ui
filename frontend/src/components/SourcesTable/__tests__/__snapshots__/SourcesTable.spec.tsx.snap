// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an empty table if no sources are provided 1`] = `
<div
  className="table-responsive"
>
  <table
    className="table sources-table table-striped table-hover"
  >
    <thead>
      <tr>
        <th>
          Indicator
        </th>
        <th>
          Updated On
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`renders correctly when the sources are updated 1`] = `
<div>
  <div
    class="table-responsive"
  >
    <table
      class="table sources-table table-striped table-hover"
    >
      <thead>
        <tr>
          <th>
            Indicator
          </th>
          <th>
            Updated On
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="table-danger"
          data-testid="sources-table-row"
        >
          <td>
            Common Reporting Standard
            <span
              class="text-uppercase"
            >
               (crs)
            </span>
          </td>
          <td>
            Sun Aug 19 2018
          </td>
        </tr>
        <tr
          class=""
          data-testid="sources-table-row"
        >
          <td>
            World Bank Indicators
            <span
              class="text-uppercase"
            >
               (wdi)
            </span>
          </td>
          <td>
            Sun Aug 19 2018
          </td>
        </tr>
        <tr
          class=""
          data-testid="sources-table-row"
        >
          <td>
            Official Development Assistance
            <span
              class="text-uppercase"
            >
               (DAC1)
            </span>
          </td>
          <td>
            Sun Aug 19 2018
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`renders correctly with default props 1`] = `
<div
  className="table-responsive"
>
  <table
    className="table sources-table table-striped table-hover"
  >
    <thead>
      <tr>
        <th>
          Indicator
        </th>
        <th>
          Updated On
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="table-danger"
        data-testid="sources-table-row"
        onClick={[Function]}
      >
        <td>
          Common Reporting Standard
          <span
            className="text-uppercase"
          >
             (crs)
          </span>
        </td>
        <td>
          Sun Aug 19 2018
        </td>
      </tr>
      <tr
        className=""
        data-testid="sources-table-row"
        onClick={[Function]}
      >
        <td>
          World Bank Indicators
          <span
            className="text-uppercase"
          >
             (wdi)
          </span>
        </td>
        <td>
          Sun Aug 19 2018
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
